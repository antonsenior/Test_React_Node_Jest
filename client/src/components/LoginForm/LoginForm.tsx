import { History } from 'history';
import * as React from 'react';
import { withRouter } from 'react-router-dom';

interface RouterProps {
  history: History;
  location: {
    hash: string;
    key: string;
    pathname: string;
    search: string;
    state: string | undefined;
  };
  match: {
    isExact: boolean;
    params: {
      id: string;
    };
    path: string;
    url: string;
  };
}

const LoginForm = (props: RouterProps) => {
  const [username, setUsername] = React.useState<string>('kcvan');
  const [password, setPassword] = React.useState<string>('Test1234!');

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label htmlFor="username">User Name</label>
        <input type="text" name="username" id="username" onChange={event => setUsername(event.target.value)} value={username} />
      </div>
      <div>
        <label htmlFor="password">Password</label>
        <input type="password" name="password" id="password" onChange={event => setPassword(event.target.value)} value={password} />
      </div>
      <button type="submit">Login</button>
    </form>
  );

  function handleSubmit(event: React.FormEvent) {
    event.preventDefault();
    const formData = {
      username,
      password
    }

    fetch('/api/login', {
      method: 'post',
        headers: {
          "Content-type": "application/json"
        },
        body: JSON.stringify(formData)
    })
    .then(response => {
      return response.json();
    })
    .then(data => {
      localStorage.setItem('myToken', data.token);
      props.history.push('/dashboard');
    })
    .catch(e => {
      console.log("Error Message: ", e.message)
    })
  }
}

export default withRouter(LoginForm);