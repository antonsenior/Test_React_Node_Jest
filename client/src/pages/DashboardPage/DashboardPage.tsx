import * as React from 'react';

const DashboardPage = (): JSX.Element => {
  return (
    <section>
      <h1>Dashboard Page</h1>
    </section>
  );
}

export default DashboardPage;