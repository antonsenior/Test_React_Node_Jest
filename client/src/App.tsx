import React from 'react';
import { BrowserRouter ,Route, Switch } from 'react-router-dom';

import DashboardPage from './pages/DashboardPage';
import Login from './pages/Login';

const App: React.FC = () => {
  return (
    <BrowserRouter>
      <section>
        <Switch>
          <Route exact path="/" component={Login} />
          <Route exact path="/login" component={Login} />
          <Route exact path="/dashboard" component={DashboardPage} />
        </Switch>
      </section>
    </BrowserRouter>
  );
}

export default App;
