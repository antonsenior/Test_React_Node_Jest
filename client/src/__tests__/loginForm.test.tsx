import React from "react";

import LoginForm from "../components/LoginForm";
import Login from "../pages/Login"
import DashboardPage from "../pages/DashboardPage"
import App from "../App";
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, render } from '@testing-library/react'
import fetchMock from 'fetch-mock'

afterEach(() => {
  fetchMock.restore();
})

describe("<LoginForm />", () => {
  test("should display empty login form by default", async () => {
    // ???
    expect(App).toBeDefined()
    const appTree = shallow(<App />)
    expect(appTree.find('LoginForm')).toBeDefined()
  });

  test("should display login Page by default", async () => {
    // ???
    expect(Login).toBeDefined()
    const tree = shallow(<Login />)
    expect(tree.find('LoginForm')).toBeDefined()
    expect(
        tree
            .find('h1')
            .text()
    ).toBe("Login")
  });

  test("should display Dashboard Page by default", async () => {
    // ???
    expect(DashboardPage).toBeDefined()
    const tree = shallow(<DashboardPage />)
    expect(tree.find('h1')).toBeDefined()
    expect(
        tree
            .find('h1')
            .text()
    ).toBe('Dashboard Page')
  });

  test("should display login form, with initial username and password by default", async () => {
    // ???
   
    expect(LoginForm).toBeDefined()
    const tree = shallow(<LoginForm />)
    expect(tree.find('form')).toBeDefined()
    expect(
        tree
            .find('form')
            .find('div')
    ).toBeDefined()
    expect(
        tree
            .find('form')
            .find('div')
            .find('label')
    ).toBeDefined()
    expect(
        tree
            .find('form')
            .find('div')
            .find('input')
    ).toBeDefined()
    expect(
        tree
            .find('form')
            .find('button')
    ).toBeDefined() 
    
    const loginComponent = render(
      <MemoryRouter>
        <LoginForm />
      </MemoryRouter>
    );
    expect(loginComponent.queryAllByText('User Name')).toHaveLength(1)
    expect(loginComponent.queryAllByText('Password')).toHaveLength(1)
    expect(loginComponent.queryAllByDisplayValue('kcvan')).toHaveLength(1)
    expect(loginComponent.queryAllByDisplayValue('Test1234!')).toHaveLength(1)
  });

  test('should return 403 error when calling login API', async () => {
    const loginComponent = render(
      <MemoryRouter>
        <LoginForm />
      </MemoryRouter>
    );

    fetchMock.post('/api/login', {
      headers: {
        "Content-type": "application/json"
      },
      status: 403
    });

    const btn = loginComponent.getByText('Login')
    fireEvent.click(btn)
  })

  test('should success when calling login API', async () => {
    const loginComponent = render(
      <MemoryRouter>
        <LoginForm />
      </MemoryRouter>
    );

    const data = {
      token: 'Token',
      success: true
    }

    fetchMock.post('/api/login', data, {
      headers: {
        "Content-type": "application/json"
      },
    });
  
    const btn = loginComponent.getByText('Login')
    const username = loginComponent.getByDisplayValue('kcvan')
    const password = loginComponent.getByDisplayValue('Test1234!')

    fireEvent.change(username, { target: { value: 'User' }})
    fireEvent.change(password, { target: { value: 'Password' }})
    await fireEvent.click(btn)
    console.log("Windows: ", document.body.textContent)
  })
});