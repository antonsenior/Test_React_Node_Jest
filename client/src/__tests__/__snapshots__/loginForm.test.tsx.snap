// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> should display login form, with initial username and password by default 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
